saleor/payment/utils.py

```diff
@@ -125,9 +125,78 @@ def clean_authorize(payment: Payment):
         raise PaymentError('Charged transactions cannot be authorized again.')
 
 
+def call_gateway(
+        func_name, transaction_kind, payment, payment_token, **extra_params):
+    """Helper that calls the passed gateway function and handles exceptions.
+
+    Additionally does validation of the returned gateway response."""
+    gateway, gateway_params = get_payment_gateway(payment.gateway)
+    gateway_response = None
+
+    #TODO: improve so only one argument is sent (dict?) instead of passing
+    # multiple kwargs, eventually keep connection params separate
+    #TODO: add additional information to payment data (e.g. shipping/billing)
+    try:
+        gateway_response = getattr(gateway, func_name)(
+            payment=model_to_dict(payment), payment_token=payment_token,
+            **extra_params, **gateway_params)
+        validate_gateway_response(gateway_response)
+    except AttributeError:
+        logger.exception('Gateway doesn\'t implement %s', func_name)
+    except GatewayError:
+        logger.exception('Gateway response validation failed')
+        gateway_response = None  # set response none as the validation failed
+    except Exception:
+        logger.exception('Gateway encountered an error')
+    finally:
+        transaction = create_transaction(
+            payment=payment,
+            kind=transaction_kind,
```

Comment:
Is this only the case with auth (in regards of capture/charge)? Cause I'd rather not make the Saleor flow dependable on the gateway. I'd rather have it the other way, so Saleor enforces some kind of flow and the gateways have to conform to it. I agree with the transaction skip though, but that can be solved with a simple additional key in the response (like 'was_authorized: True/False') that could then create an additional auth transaction.

Reply:
As far as I know this is the only case.

An additional key may not be enough. If it is truly a separation operation in third-party gateway, we should record all the info includes raw response, amount, currency, etc.

Reply:
This means that returning a list might be unavoidable.. Kinda sad if it comes to that, I like how simple it is, but this will make it more complex just to conform to that one case of two operations. I wonder if there's a better solution.

Reply:
I know this looks not good. But recording all the information to a payment is more important.

Maybe we could consider supporting both a list or a dict? For normal case, ```call_gateway``` will take a dict as returned value, but if it is a list, the ```call_gateway``` should also be able to handle it.

Reply:
I think I'd prefer at least enforcing a list on the Saleor side. Sure, the gateway can return a dict for simplicity, but on Saleor I'd have it made a list if it's a dict to ensure that we have one logic (of say iterating and creating transactions). I'll go with this, I'm a bit sceptical having the gateway return what kind of transaction it did, but sure, let's go for it :)

Reply:
That's better. At least mypy could check the types easily.

Maybe it worth considering let the gateway itself returning the transaction kind and use it directly in creating transactions.